File: recipes/subms-otel/rust/src/tracing_observer.rs
//! Span-emitting observer. On every [`SubMsObserver::on_record`] this fires a
//! tiny span whose start time is reconstructed as `now - ns`; if the calling
//! thread already has an active OTEL Context with a parent span, that parent
//! is inherited so a higher-level trace (e.g. a Spring HTTP filter) can flow
//! through the bench instrumentation.

use std::time::SystemTime;

use opentelemetry::Context;
use opentelemetry::trace::{Span, SpanBuilder, Tracer};

use subms::{ObservationCtx, SubMsBenchSummary, SubMsObserver};

use crate::bridge::attributes_from_ctx;

/// Span name used for every emitted record. Stage identity is on the attrs.
pub const TRACING_SPAN_NAME: &str = "subms.stage.record";

/// Span-emitting observer. Holds a tracer + the span name template.
pub struct TracingObserver<T: Tracer + Send + Sync> {
    tracer: T,
}

impl<T> TracingObserver<T>
where
    T: Tracer + Send + Sync,
    T::Span: Span + Send + Sync + 'static,
{
    /// Build bound to the given tracer.
    pub fn new(tracer: T) -> Self {
        Self { tracer }
    }
}

impl<T> SubMsObserver for TracingObserver<T>
where
    T: Tracer + Send + Sync,
    T::Span: Span + Send + Sync + 'static,
{
    fn on_record(&self, ctx: &ObservationCtx, ns: u64) {
        let now = SystemTime::now();
        let start = now
            .checked_sub(std::time::Duration::from_nanos(ns))
            .unwrap_or(now);
        let parent_ctx = Context::current();
        let attrs = attributes_from_ctx(ctx);
        let mut span = SpanBuilder::from_name(TRACING_SPAN_NAME)
            .with_start_time(start)
            .with_attributes(attrs)
            .start_with_context(&self.tracer, &parent_ctx);
        span.end_with_timestamp(now);
    }

    fn on_summarize(&self, _summary: &SubMsBenchSummary) {
        // No span on summarize - the bridge's export_timer / export_summary
        // already covers the post-bench narrative.
    }
}

#[cfg(test)]
#[path = "tracing_observer_tests.rs"]
mod tracing_observer_tests;

File: recipes/subms-otel/java/src/main/java/com/submillisecond/recipes/otel/tracing/TracingObserver.java
package com.submillisecond.recipes.otel.tracing;

import com.submillisecond.recipes.otel.SubMsOtel;
import com.submillisecond.perf.SubMsBenchSummary;
import com.submillisecond.perf.SubMsObservationCtx;
import com.submillisecond.perf.SubMsObserver;
import io.opentelemetry.api.common.Attributes;
import io.opentelemetry.api.trace.Span;
import io.opentelemetry.api.trace.SpanBuilder;
import io.opentelemetry.api.trace.SpanContext;
import io.opentelemetry.api.trace.Tracer;
import io.opentelemetry.context.Context;

import java.time.Instant;
import java.util.Objects;

/**
 * {@link SubMsObserver} that emits a short OTEL span per {@code onRecord} call.
 *
 * <p>Span start is reconstructed as {@code now - ns} so the duration matches the harness's
 * measured nanoseconds. If a calling thread already has an active span (e.g. a Spring HTTP
 * filter started a trace), the {@link Context#current()} parent is inherited; otherwise the
 * span is rooted.
 *
 * <p>{@link #onSummarize(SubMsBenchSummary)} is intentionally a no-op - the post-bench
 * narrative is already covered by
 * {@link SubMsOtel#exportTimer(com.submillisecond.perf.SubMsTimer, Tracer)} +
 * {@link SubMsOtel#exportSummary(SubMsBenchSummary, io.opentelemetry.api.metrics.Meter)}.
 */
public final class TracingObserver implements SubMsObserver {

    /** Span name used for every emitted record. Stage identity is on attrs. */
    public static final String TRACING_SPAN_NAME = "subms.stage.record";

    private final Tracer tracer;

    public TracingObserver(Tracer tracer) {
        this.tracer = Objects.requireNonNull(tracer, "tracer");
    }

    @Override
    public void onRecord(SubMsObservationCtx ctx, long ns) {
        Instant end = Instant.now();
        Instant start = end.minusNanos(ns);
        Attributes attrs = SubMsOtel.ctxAttributes(
                ctx.workload(), ctx.lang(), ctx.stage(), ctx.stageKind());
        Context parentCtx = Context.current();
        SpanBuilder b = tracer.spanBuilder(TRACING_SPAN_NAME)
                .setParent(parentCtx)
                .setStartTimestamp(start)
                .setAllAttributes(attrs);
        Span span = b.startSpan();
        SpanContext sc = span.getSpanContext();
        if (sc != null && sc.isValid()) {
            // touch to prevent the unused-var warning across surfaces that don't read it
        }
        span.end(end);
    }

    @Override
    public void onSummarize(SubMsBenchSummary summary) {
        // intentional no-op
    }
}
